File: CAPA_NEGOCIO/Reports/Models/EstudiantesUpdateReportModel.cs
using APPCORE;
using CAPA_NEGOCIO.UpdateModule.Model;
using DataBaseModel;

namespace CAPA_NEGOCIO.Reports.Models
{
    public class EstudiantesUpdateReportModel : QueryClass
    {
        public int? Id { get; set; }
        public string? Id_familia { get; set; }
        public string? Codigo { get; set; }
        public string? Primer_nombre { get; set; }
        public string? Segundo_nombre { get; set; }
        public string? Primer_apellido { get; set; }
        public string? Segundo_apellido { get; set; }
        public DateTime? Fecha_nacimiento { get; set; }
        public string? Periodo_Lectivo_Update { get; set; }
        public string? Grado { get; set; }
        public string? Seccion { get; set; }
        public string? Nivel { get; set; }
        public int? Periodo_lectivo_id { get; set; }
        public string? Nombre_completo { get { return $"{Primer_nombre} {Segundo_nombre} {Primer_apellido} {Segundo_apellido}"; } }


        [JsonProp]
        public SecurityOption? SecurityOption { get; set; }

        public override string GetQuery()
        {
            var filter = filterData?.Find(f => f.PropName == "Periodo_Lectivo_Update");
            if (Periodo_Lectivo_Update == null && filter == null)
            {
                var PeriodoActivo = Periodo_lectivos.PeriodoActivo();              
                Periodo_Lectivo_Update = PeriodoActivo?.Nombre_corto;
            } else if(Periodo_Lectivo_Update == null && filter != null)
            {
                 Periodo_Lectivo_Update = filter?.Values.FirstOrDefault();
            }
            var periodo = new Periodo_lectivos().Find<Periodo_lectivos>(
                FilterData.Equal("Nombre_corto", Periodo_Lectivo_Update)
            );
            Periodo_lectivo_id = periodo?.Id;
            return $@"SELECT
                e.id_familia,
                e.codigo,
                e.id,
                e.primer_nombre,
                e.segundo_nombre,
                e.primer_apellido,
                e.segundo_apellido,	
                e.fecha_nacimiento,
                e.SecurityOption,
                e.Periodo_Lectivo_Update,
                c.grado,
                s.nombre as seccion,
                n.nombre as nivel,
                ec.periodo_lectivo_id
            FROM
                update_data.estudiantes_data_update as e
            INNER JOIN estudiante_clases as ec ON
                e.id = ec.estudiante_id
            INNER JOIN clases as c ON
                c.id = ec.clase_id
            INNER JOIN secciones AS s ON
                s.id = ec.seccion_id
            INNER JOIN niveles AS n ON
                n.id = c.nivel_id
            WHERE
                e.Periodo_Lectivo_Update = '{Periodo_Lectivo_Update}'
                AND ec.periodo_lectivo_id = {Periodo_lectivo_id}
            Order by grado";
        }
    }
}

File: UI/wwwroot/Reportes/Model/EstudiantesUpdateReportModel.js
//@ts-check
import { Periodo_lectivos } from '../../Model/Periodo_lectivos.js';
// @ts-ignore
import { ModelProperty } from '../../WDevCore/WModules/CommonModel.js';
import { EntityClass } from '../../WDevCore/WModules/EntityClass.js';
import { SecurityOption, SecurityOption_ModelComponent, securityOptions } from '../../update/Model/Estudiantes_ModelComponent.js'; // Importar SecurityOption de la clase de referencia


/**
 * @class EstudiantesUpdateReportModel
 * @extends EntityClass
 * @description Modelo de datos para la actualización de estudiantes, utilizado para consultas de solo lectura.
 */
class EstudiantesUpdateReportModel extends EntityClass {

    /**
     * @param {Partial<EstudiantesUpdateReportModel>} [props]
     */
    constructor(props) {
        super(props, "Reportes");
        /** @type {number | null} */ this.Id = null;
        /** @type {string | null} */ this.Id_familia = null;
        /** @type {string | null} */ this.Idtfamilia = null;
        /** @type {string | null} */ this.Codigo = null;
        /** @type {string | null} */ this.Nombre_completo = null;
        /** @type {string | null} */ this.Primer_nombre = null;
        /** @type {string | null} */ this.Segundo_nombre = null;
        /** @type {string | null} */ this.Primer_apellido = null;
        /** @type {string | null} */ this.Segundo_apellido = null;
        /** @type {Date | null} */ this.Fecha_nacimiento = null; // DateTime? en C# se mapea a Date | null en JS
        /** @type {string | null} */ this.Periodo_Lectivo_Update = null;
        /** @type {string | null} */ this.Grado = null;
        /** @type {string | null} */ this.NombreGrado = null;
        /** @type {string | null} */ this.Seccion = null;
        /** @type {string | null} */ this.Nivel = null;
        /** @type {number | null} */ this.Periodo_lectivo_id = null;
        /** @type {SecurityOption | null} */ this.SecurityOption = new SecurityOption(); // Asumiendo que SecurityOption es una clase o un objeto
       
        // Aplicar propiedades pasadas en el constructor
        if (props) {
            /** @type {keyof EstudiantesUpdateReportModel} */
            let prop;
            for (prop in props) {
                if (Object.prototype.hasOwnProperty.call(props, prop)) {
                    // @ts-ignore: TypeScript no puede inferir completamente que prop es una clave válida aquí
                    this[prop] = props[prop];
                }
            }
        }
        /** @type {string | null} */ this.SecurityOptionLabel = this.SecurityOption?.Descripcion ?? "";
    }
    /**
    * @returns {Promise<any>}
    */
    async ReenviarBoleta() {
        return await this.GetData("ApiReportes/GetEstudiantesConRecorridos");
    }
}

export { EstudiantesUpdateReportModel };

class EstudiantesUpdateReportModel_Modelcomponent {
    /** @param {Partial<EstudiantesUpdateReportModel_Modelcomponent>} [props] */
    constructor(props) {
        //super(props, 'Reportes'); // 'EstudiantesUpdate' como nombre de la entidad para la API
        //@ts-ignore   
        for (const prop in props) { this[prop] = props[prop]; };
        //this.Model = new EstudiantesUpdateReportModel(); // Mantener una instancia del modelo de datos
    }

    /**@type {ModelProperty}*/ Id = { type: "number", primary: true, hidden: true, hiddenFilter: true };
    //**@type {ModelProperty}*/ Id_familia = { type: "text", label: "ID Familia", hiddenFilter: true };
    /**@type {ModelProperty}*/ Idtfamilia = { type: "text", label: "ID Familia", hiddenFilter: true };
    /**@type {ModelProperty}*/ Codigo = { type: "text", label: "Código", hiddenFilter: true };
    /**@type {ModelProperty}*/ Nombre_completo = { type: "text", hiddenFilter: true };
    //**@type {ModelProperty}*/ Primer_nombre = { type: "text", label: "Primer Nombre" , hiddenFilter: true};
    //**@type {ModelProperty}*/ Segundo_nombre = { type: "text", label: "Segundo Nombre", require: false, hiddenFilter: true };
    //**@type {ModelProperty}*/ Primer_apellido = { type: "text", label: "Primer Apellido" , hiddenFilter: true};
    //**@type {ModelProperty}*/ Segundo_apellido = { type: "text", label: "Segundo Apellido", require: false, hiddenFilter: true };
    /**@type {ModelProperty}*/ Fecha_nacimiento = { type: "date", label: "Fecha de Nacimiento", hiddenFilter: true };
    /**@type {ModelProperty}*/ Periodo_Lectivo_Update = { type: "select", hiddenInTable: true, label: "Periodo de Actualización", Dataset: [] };
    /**@type {ModelProperty}*/ NombreGrado = { type: "text", label: "Grado", hiddenFilter: true };
    /**@type {ModelProperty}*/ Seccion = { type: "text", label: "Sección", hiddenFilter: true };
    /**@type {ModelProperty}*/ Nivel = { type: "text", label: "Nivel", hiddenFilter: true };
    /**@type {ModelProperty}*/ Periodo_lectivo_id = { type: "number", label: "ID Periodo Lectivo", hidden: true, hiddenFilter: true };

    /**@type {ModelProperty}*/ SecurityOptionLabel = {
        type: 'WRADIO', label: "Seguro estudiantil",
        Dataset: securityOptions,
        ModelObject: new SecurityOption_ModelComponent(), // Usar SecurityOption_ModelComponent para la UI
        require: false
    };

    // El método GetQuery ya no es necesario aquí, ya que EntityClass maneja la interacción con la API
    // a través de las propiedades y el nombre de la entidad.
}

export { EstudiantesUpdateReportModel_Modelcomponent };

